src/julia.h

```diff
@@ -1358,6 +1372,50 @@ extern DLLEXPORT jl_compileropts_t jl_compileropts;
 #define JL_COMPILEROPT_DUMPBITCODE_ON 1
 #define JL_COMPILEROPT_DUMPBITCODE_OFF 2
 
+DLLEXPORT void gc_queue_root(void *root);
+void gc_queue_binding(void *bnd);
+void gc_setmark_buf(void *buf, int);
+DLLEXPORT void gc_wb_slow(void* parent, void* ptr);
+
+static inline void gc_wb_binding(void *bnd, void *val)
+{
+    if (__unlikely((*(uintptr_t*)bnd & 1) == 1 && (*(uintptr_t*)val & 1) == 0))
+        gc_queue_binding(bnd);
+}
+
+static inline void gc_wb_fwd(void* parent, void* ptr)
+{
+    // if parent is marked and ptr is clean
+    if(__unlikely((*((uintptr_t*)parent) & 1) == 1 && (*((uintptr_t*)ptr) & 1) == 0)) {
+        // the set lsb indicates this object must stay in the remset until the next
+        // long collection
+        gc_queue_root((void*)((uintptr_t)ptr | 1));
+    }
+}
+
+static inline void gc_wb(void *parent, void *ptr)
+{
+    if (__unlikely((*((uintptr_t*)parent) & 1) == 1 &&
+                   (*((uintptr_t*)ptr) & 1) == 0))
+        gc_queue_root(parent);
+}
+
+static inline void gc_wb_buf(void *parent, void *bufptr)
+{
+    // if parent is marked and buf is not
+    if (__unlikely((*((uintptr_t*)parent) & 1) == 1))
+                   //                   (*((uintptr_t*)bufptr) & 3) != 1))
+          gc_setmark_buf(bufptr, *(uintptr_t*)parent & 3);
```

Comment:
It does not include what is referred in the code as the age bit which is where we remember if the object already survived one collection. Once an object turns old for real (i.e., after 2 survivals) we mark it in its low bits (marked as 01 instead of 11). The transition happens at this point and we check for child objects/buffers also there.